### DISCUSSION TITLE
open MPI on linux

### DISCUSSION THREAD

**Discussion by danFrodo** [OpenRadioss/OpenRadioss#1484] [Users]
I am running openradioss on an ubuntu 20.04 install.  LSCPU commands shows the following:

Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         46 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  36
  On-line CPU(s) list:   0-35
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Xeon(R) Gold 6254 CPU @ 3.10GHz
    CPU family:          6
    Model:               85
    Thread(s) per core:  2
    Core(s) per socket:  18
    Socket(s):           1

From this I am thinking I should be able to run -np 18 with THREADS=2.  But that doesn't work it seems.  For -np, I use the following command:
mpiexec -n 9  --map-by socket:PE=$OMP_NUM_THREADS --bind-to core /home/APPS/ALTAIR/OPENRADIOSS/OpenRadioss/exec/engine_linux64_gf_ompi -i  arm100B_0001.rad

What works:
440K element model
- -nt 36 (run time 926 sec)
- -np 18 THREADS=1 (run time 468 sec)
- -np 9  THREADS=2 (run time 465 sec)

It does not work if -np 18 &  OMP_NUM_THREADS is set to 2, using the above command.  

But I can run with simplified mpiexec:
mpiexec -n 18  /home/APPS/ALTAIR/OPENRADIOSS/OpenRadioss/exec/engine_linux64_gf_ompi -i  arm100C_0001.rad

which runs same time as above, at 459 sec
 NUMBER OF SPMD DOMAINS          18
 NUMBER OF THREADS PER DOMAIN     2
 NUMBER OF HMPP PROCESSES        36

This run seems to be using the domains & threads but is same speed as the ones using the more extensive mpiexec command.

I am wondering if I am not getting the most computing power that I have available. 

Anyone have any thoughts?  

Thanks,
Dan

-  
-

**Comment by elequiniou** [comment]
Dear Dan, 

Thanks a lot for your interest in OpenRadioss and for taking the time to precisely describe your issue. Everything looks OK to me. Indeed you only have 18 physical cores. Each core is multi-threaded giving the impression to have 36 CPUs. For some applications, using so-called Hyper-Threading could help, but for OpenRadioss, this is marginal. The maximum performance is already achieved with 18 processes on 18 cores (it could help more for application less compute and memory intensive). So, your setting with 18 MPI (or 9 MPI x 2 OMP) is already optimal to get the best of your platform!

Thanks,

Eric.

**Comment by danFrodo** [comment]
Thanks very much Eric, this is very helpful!  I've been using Radioss for a
couple years now.  I needed it to use a leg biomechanical model from Univ
Marseilles (amazing model).  I had Abaqus explicit in-house which I had
used for many years but once I got good at Radioss I switched over.  Now in
new job in robotics & using mix of altair radioss & open radioss.  I've
worked w Marian B a lot in the past.

By the way, do you know of any documentation on connectors or using
spring-Beam or other for connectors?  I haven't used those much but need
them now.  The Radioss manual has basic info but wondering if there's more
info.  I need to model actuators & bearings at robot joints.

Check out Agility Robotics and you'll see what I am trying to model.  My
first thing is looking at robot falling events.

Dan

**Comment by MarianBulla** [comment]
Hi Dan,
for modelling of connections, there are many possibilities in Radioss (and OpenRadioss):
- Using contact interfaces e.g. TYPE2 (tied with possibility of releasing) or TYPE10 (tied after contact) ...
- Cohesive zone elements: modelling of adhesives and/or spotwelds, rivets, ...
- 1D-elements e.g. spring TYPE13 (most flexibel, but passive) spring TYPE46 (active element)
- 3D-elements for modelling of prestresed screws
.
. ...and many other things
So, it depends on your application.

We are enhancing our Examples section, here at GitHub, with those small examples.

So stay tuned :-)

Thanks and best reagards,
Marian


Hope it helps you ?

**Comment by TessellateDataScience** [comment]
Hiya community, I did some brief testing of hardware parallelisation for vehicle crash dynamics of a mesh size of ~ 500K elements [*]. Note this means I allocated more cores (`-np`), which invokes MPI.

Seems parallelisation efficiency is pretty good until reaching the maximum number of cores we have installed (see table below).

| numbCores | Tcomp [s] | scaleUp [**] |
| --------- | --------- | -------------|
|  		  1 | 	   3279 | 			 - |
|  		  2 | 	   1807 |         1.84 |
|  		  4 |		882 |		  2.04 |
|  		  8 |		501	|		  1.73 |
| 		 16 |		301	|		  1.66 |
| 		 32 |		192 |		  1.56 |

[* ] Checked on a HP-DL380-Gen9 (2 * Xeon CPUs), running OpenRadioss on Ubuntu 22 LTS.

[**] ScaleUp is improvement ratio from previous number of cores used during computation (numbCores_i-1 / numbCores_i). Values are estimated at NC = 500 (5th iteration).

**Comment by CharlieM-57** [comment]
Hello Y'all
I am new at this. How do I determine how many cores and threads my computer has?
Attached is a screen shot about my linux computer.
Thanks

![about the computer 24Mar26](https://github.com/user-attachments/assets/159accff-5be0-4bb5-8702-df31d331ee6c)


Charlie

**Reply by TessellateDataScience** [reply]
Heya, Google AI tells me the i7-4770 is a desktop processor (released 2013) featuring 4 cores, 8 threads, and a 3.4 GHz base clock (up to 3.9 GHz boost), which can be verified via Intel's website (specifications): www.intel.com/content/www/us/en/products/sku/75122/intel-core-i74770-processor-8m-cache-up-to-3-90-ghz/specifications.html

You're wanting more cores (over more threads), I believe.

**Comment by CharlieM-57** [comment]
Hi Y'all,  I am trying to set up and run OpenRadioss with limited Linux Ubuntu language abilities. 
I think I successfully downloaded and compiled OpenRadioss and openmpi_4.1.2 without blowing up my computer or injuring myself.
I tried to download Visual Studio via flatpak but there was some issue and the program was locked in a container(?).
I am trying to download and set-up OpenRadioss_gui but am not having any success.
My input will be prepared by FreeCAD v1.1 which I can use without much issue. So I will need some tools to convert files from there into something that OpenRadioss can read.
Any help will be appreciated.
Thanks,
Charlie

**Reply by TessellateDataScience** [reply]
Hi Charlie, good you didn't need protective equipment for install. If youre looking to 'kick the tires' OR eventually transfer a model from your local computer to HPC (university cluster), I'm a convert to using Docker. You don't actually need the GUI to use ORadioss, you can launch from the command line (Terminal). 

If you don't feel like rolling your own (basically reusing your install commands), I have a pretty recent version that you're welcome to use. To get going, launch the Terminal app, then:
```
# install Docker: https://docs.docker.com/engine/install/ubuntu/
...

# get my Docker (prebuilt) 'image'
docker image pull nchowlett/openradioss:0.3

# navigate to upload directory (local computer)
mkdir ORFiles
cd ORFiles

# boot this (image) environment
docker run -it --mount "type=bind,src=$pwd,target=/OpenRadioss/data" nchowlett/openradioss:0.3 sh

# navigate into working directory (remote environment)
cd /OpenRadioss/data
````

Then you'll need to configure the environment ready for computation by saving the following code into  _caseRun.sh_ in the ORFiles dir.:
```
export OPENRADIOSS_PATH=/OpenRadioss
export RAD_CFG_PATH=$OPENRADIOSS_PATH/hm_cfg_files
export RAD_H3D_PATH=$OPENRADIOSS_PATH/extlib/h3d/lib/linux64
export LD_LIBRARY_PATH=$OPENRADIOSS_PATH/extlib/hm_reader/linux64:/opt/openmpi/lib:$LD_LIBRARY_PATH
export PATH=/opt/openmpi/bin:$PATH
export OMP_STACKSIZE=400m
export OMP_NUM_THREADS=1

myjobname='radiossTest'
rm radiossTestA00*
rm radiossTest_*.vtk

echo "Running Starter:"
/OpenRadioss/exec/starter_linux64_gf -i $myjobname"_0000.rad" -np 4
```
Finally, check your environment is computational ready:
```
# ensure slack permissions (execute only once)
chmod 777 bashFile.sh
chown root:root bashFile.sh

# run check of OpenRadioss
./caseRun.sh
```

If you get output from OpenRadioss printed to your screen then hopefully everything is ready (OpenRadioss is at least running):
```
************************************************************************
**                                                                    **
**                                                                    **
**                        OpenRadioss Starter                         **
**                                                                    **
**            Non-linear Finite Element Analysis Software             **
```

Any install issues feel free to hit me up. Note there might be performance sacrifices, but reproducibility is golden for me.

You can find numerous [OpenRadioss source files](https://openradioss.org/models/) online, I'm assuming they have been tested for the version of OpenRadioss I have installed within Docker, but I haven't checked.

**Reply by CharlieM-57** [reply]
G'Day. Many thanks for your help. I will have to read up on this over the weekend. My programming skills were started with punch cards and DOS3.1. Not quite Fred Flinstone, but no experience since then. 
Do I need Visual Studios which is what one of the YouTubers said was needed when installing OpenRadioss. I found that you can download a copy from flatpak but when I did, there were some warnings that came with it, like containerization...
According to the YouTuber, Dominik Kern Dynamics (17March 2026) about OpenRadioss FEM: Preprocessing with FreeCAD on Linux, I can use FreeCAD. (www.youtube.com/watch?v=ZDYEC7Wtsrw)
I would prefer to use FreeCAD because that is the only software I have access to. Even though I am a "Visiting Faculty" in the Chemistry dept., it is an unpaid position. I get access to the university library, I help my professor friend, and I am allowed to torture students. If I had access to the FEM software in Mechanical Engineering, I would have solved the problem by now. 
Anyway. Dominik says to create geometry and mesh (Gmsh) in FreeCAD, export in Abaqus format, edit, then run OpenRadioss tool inp2rad. Downloading / setup of inp2rad was not successful. I will try again over the weekend. 
I would prefer to use the GUI since it could be simpler until I get fluent in Ubuntu.
I have successfully downloaded ParaView, without injury, and now need to figure out how to use it. I need to again figure out and download the OpenRadioss tools anim_to_vtk and th_to_csv programs. 

Would Docker be able to help with all this?
Have a good long weekend.
Charlie

**Reply by TessellateDataScience** [reply]
Docker is probably not the way to go if you wanted to visually inspect (or primarily use GUI-based software). However, I reckon you're torturing yourself unnecessarily if you want to compile from source (I build the Docker image to partially save others the lower-level coding + risk of throwing said computer out the window that I experienced). If you still want to compile then you'll probably need a _text editor_, but doesn't have to be VStudio.

You should be able to get access to Coreform Cubit for free (non-commercial application assumed). I use it to generate hex-based meshes that Gmsh (also free to use) needs extra work to generate. Also, a tidy Python script to [convert LSDyna-to-Radioss format](https://github.com/luchete80/radioss_tutorials/blob/master/lsdyna2rad/convert_orig.py) is a bonus if you favour Cubit (I also couldn't get inp2rad to work).

In summary I'd:
1. Install your GUI (pre/post-processor) software (such as FreeCAD, Gmsh, Cubit, ParaView) in a [_virtual machine_](https://www.virtualbox.org/manual/topics/installation.html). This allows you to distribute it to others that have their own laptops, fairly easily.
2. Use my OpenRadioss (Docker) environment as your (computing) virtual machine. I'm actively maintaining it as I'm using it for computations also.

Regards, Nick
